### ISSUE TITLE
"Install locally with LXD" Guide lacks a workaround for an avahi issue with privileged containers and rlimits

### ISSUE THREAD

**Issue by dshcherb** [CanonicalLtd/maas-docs#209]
TL;DR: as it is recommended to use a privileged container, a separate user namespace is not created for a container which leads to an error related to rlimits for the avahi-daemon if two identical uids are used from the host's point of view and the container's point of view to mean different users.

This requires an update:
https://github.com/CanonicalLtd/maas-docs/blob/master/en/installconfig-lxd-install.md

The long version:

During the MAAS installation process the following two packages get installed:
`avahi-daemon avahi-utils`

The service fails as follows:

```-- Unit avahi-daemon.service has begun starting up.
янв 06 23:49:16 maas avahi-daemon[24036]: Found user 'avahi' (**UID 113**) and group 'avahi' (GID 118).
янв 06 23:49:16 maas avahi-daemon[24036]: Successfully dropped root privileges.
янв 06 23:49:16 maas avahi-daemon[24036]: chroot.c: fork() failed: Resource temporarily unavailable
янв 06 23:49:16 maas avahi-daemon[24036]: failed to start chroot() helper daemon.
янв 06 23:49:16 maas systemd[1]: avahi-daemon.service: Main process exited, code=exited, status=255/n/a
янв 06 23:49:16 maas systemd[1]: Failed to start Avahi mDNS/DNS-SD Stack.
```

Which is not descriptive at all (note UID **113**):

A simple reproducer:
```
root@maas:~# strace avahi-daemon
setrlimit(RLIMIT_CORE, {rlim_cur=0, rlim_max=0}) = 0
setrlimit(RLIMIT_DATA, {rlim_cur=4096*1024, rlim_max=4096*1024}) = 0
setrlimit(RLIMIT_FSIZE, {rlim_cur=0, rlim_max=0}) = 0
setrlimit(RLIMIT_NOFILE, {rlim_cur=768, rlim_max=768}) = 0
setrlimit(RLIMIT_STACK, {rlim_cur=4096*1024, rlim_max=4096*1024}) = 0
**setrlimit(RLIMIT_NPROC, {rlim_cur=3, rlim_max=3}) = 0**
setrlimit(RLIMIT_MEMLOCK, {rlim_cur=0, rlim_max=0}) = 0
...
clone(child_stack=0, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7fc7e8d55b50) = -1 EAGAIN (Resource temporarily unavailable)
close(5) = 0
close(6) = 0
write(2, "chroot.c: fork() failed: Resourc"..., 57chroot.c: fork() failed: **Resource temporarily unavailable**) = 57
```

This line in particular sets RLIMIT_NPROC:

`setrlimit(RLIMIT_NPROC, {rlim_cur=3, rlim_max=3}) = 0`

man setrlimit says that this is a global limit for a particular UID (that is kernel-wide for this uid - not for child processes only):

> RLIMIT_NPROC
> The maximum number of processes (or, more precisely on Linux, threads) that can be created **for the real user ID of the calling process**. **Upon encountering this limit**, **fork(2) fails with the error EAGAIN**.
> This limit is not enforced for processes that have either the CAP_SYS_ADMIN or the CAP_SYS_RESOURCE capability.

/etc/avahi/avahi-daemon.conf:
```
[rlimits]
#rlimit-as=
rlimit-core=0
rlimit-data=4194304
rlimit-fsize=0
rlimit-nofile=768
rlimit-stack=4194304
rlimit-nproc=3
```

We care about this as the same uid may be used both "inside" and "outside" of a container:

```
# the host
$ cut -d: -f1,3 /etc/passwd | grep -i 113
colord:113

# a privileged container
root@maas:~# cut -d: -f1,3 /etc/passwd | grep -i 113
avahi:113
```

On the host there are a lot of processes already with that uid so a limit of 3 set somewhere in a container but system-wide will most likely ruin things:

```
$ ps -ef | grep -i colord | tail -n4
colord   27560 26300  1 янв05 ?     00:39:31 /usr/bin/python /usr/bin/neutron-server --config-file=/etc/neutron/neutron.conf --config-file=/etc/neutron/plugins/ml2/ml2_conf.ini --log-file=/var/log/neutron/neutron-server.log
colord   27561 26300  1 янв05 ?     00:38:41 /usr/bin/python /usr/bin/neutron-server --config-file=/etc/neutron/neutron.conf --config-file=/etc/neutron/plugins/ml2/ml2_conf.ini --log-file=/var/log/neutron/neutron-server.log
colord   30378  9620  0 00:12 ?        00:00:00 avahi-daemon: running [maas.local]
colord   30401 30378  0 00:12 ?        00:00:00 avahi-daemon: chroot helper
```

A simple proof of concept for a workaround:

```
root@maas:~# avahi-daemon  --no-rlimits
Found user 'avahi' (UID 113) and group 'avahi' (GID 118).
Successfully dropped root privileges.
avahi-daemon 0.6.32-rc starting up.
...
```

```
 lxc info maas | grep Pid
Pid: 9620

lxc profile show maas | grep priv
  security.privileged: "true"
```

With privileged containers a separate user namespace is not created which is why we care about this:

```
 ps -o pid,ppid,uid,user,cmd -p 9620 -p 9508 --ppid 9620 --forest
  PID **PPID** UID USER CMD
 **9508** 1 0 root [lxc monitor] /var/lib/lxd/containers maas
 **9620** **9508** 0 root \_ /sbin/init
 9886 9620 0 root \_ /lib/systemd/systemd-journald
```

Both point to the same one (user:[4026531837]):
```
$ sudo ls -l /proc/9620/ns/
total 0
lrwxrwxrwx 1 root root 0 янв 6 13:13 cgroup -> cgroup:[4026534706]
lrwxrwxrwx 1 root root 0 янв 6 13:13 ipc -> ipc:[4026534637]
lrwxrwxrwx 1 root root 0 янв 6 13:13 mnt -> mnt:[4026534561]
lrwxrwxrwx 1 root root 0 янв 6 13:13 net -> net:[4026534640]
lrwxrwxrwx 1 root root 0 янв 6 13:13 pid -> pid:[4026534638]
lrwxrwxrwx 1 root root 0 янв 6 13:13 user -> user:[**4026531837**]
lrwxrwxrwx 1 root root 0 янв 6 13:13 uts -> uts:[4026534634]
$ ps 9620
  PID TTY STAT TIME COMMAND
 9620 ? Ss 0:13 /sbin/init
$ ps -O ppid 9620
  PID PPID S TTY TIME COMMAND
 9620 9508 S ? 00:00:13 /sbin/init
$ sudo ls -l /proc/9508/ns/
total 0
lrwxrwxrwx 1 root root 0 янв 6 23:22 cgroup -> cgroup:[4026531835]
lrwxrwxrwx 1 root root 0 янв 6 23:22 ipc -> ipc:[4026531839]
lrwxrwxrwx 1 root root 0 янв 6 23:22 mnt -> mnt:[4026531840]
lrwxrwxrwx 1 root root 0 янв 6 23:22 net -> net:[4026531969]
lrwxrwxrwx 1 root root 0 янв 6 23:22 pid -> pid:[4026531836]
lrwxrwxrwx 1 root root 0 янв 6 23:22 user -> user:[**4026531837**]
lrwxrwxrwx 1 root root 0 янв 6 23:22 uts -> uts:[4026531838]
```

After commenting out rlimit configuration in the /etc/avahi/avahi-daemon.conf:

```
root@maas:~# systemctl start avahi-daemon.service 
root@maas:~# systemctl status avahi-daemon.service 
● avahi-daemon.service - Avahi mDNS/DNS-SD Stack
   Loaded: loaded (/lib/systemd/system/avahi-daemon.service; enabled; vendor preset: enabled)
   Active: active (running) since Сб 2017-01-07 00:12:57 MSK; 2s ago
 Main PID: 27229 (avahi-daemon)
   Status: "avahi-daemon 0.6.32-rc starting up."
    Tasks: 2
   Memory: 1.8M
      CPU: 27ms
   CGroup: /system.slice/avahi-daemon.service
           ├─27229 avahi-daemon: running [maas.local
           └─27230 avahi-daemon: chroot helpe

янв 07 00:12:57 maas avahi-daemon[27229]: Joining mDNS multicast group on interface eth0.IPv6 with address fe80::216:3eff:fea3:4b1e.
янв 07 00:12:57 maas avahi-daemon[27229]: New relevant interface eth0.IPv6 for mDNS.
янв 07 00:12:57 maas avahi-daemon[27229]: Joining mDNS multicast group on interface eth0.IPv4 with address 10.10.10.58.
янв 07 00:12:57 maas avahi-daemon[27229]: New relevant interface eth0.IPv4 for mDNS.
янв 07 00:12:57 maas avahi-daemon[27229]: Network interface enumeration completed.
янв 07 00:12:57 maas avahi-daemon[27229]: Registering new address record for fe80::216:3eff:fecd:6a80 on eth1.*.
янв 07 00:12:57 maas avahi-daemon[27229]: Registering new address record for 10.10.101.2 on eth1.IPv4.
янв 07 00:12:57 maas avahi-daemon[27229]: Registering new address record for fe80::216:3eff:fea3:4b1e on eth0.*.
янв 07 00:12:57 maas avahi-daemon[27229]: Registering new address record for 10.10.10.58 on eth0.IPv4.
янв 07 00:12:58 maas avahi-daemon[27229]: Server startup complete. Host name is maas.local. Local service cookie is 3661524336.
```

It is an open question as to what to recommend in this case but a workaround should be mentioned at least.

**Comment by pmatulis** [comment]
@dshcherb Thanks a lot for this detailed bug report. :ribbon:

**Comment by dshcherb** [comment]
@pmatulis 

Let me know if you need any more info: this will probably not fire on a fresh ubuntu installation with a single container - there simply won't be two UIDs used for different purposes and the avahi-daemon won't run into this.

A simple way to reproduce this could be:

1) Create a privileged container for MAAS deployment;
2) Install maas there and check which uid does the avahi user have (in my case it was 113);
3) Stop the avahi daemon;
4) Create a bunch of processes with that uid on the host;
E.g.:
```
cp `which cat` ./
sudo chown 113 ./cat
# setuid
sudo chmod u+s ./cat
./cat &
[3] 31807
ps -o uid 31807
  UID
  113
...

```
6) Start the avahi daemon in the container
